alg.py

```diff
@@ -1,182 +1,367 @@
+        prio        prio[i] is the priority for task i
+        dead        dead[i] is the deadline for task i
+
+        """
+        # Check for correct inputs
+        if not isinstance(num_tasks, int) or num_tasks < 0:
+            raise ValueError("num_tasks should be a positive integer")
+        if not len(stages) == num_tasks:
+            raise ValueError("stages should have length num_tasks")
+        if not len(time) == num_tasks:
+            raise ValueError("time should have length num_tasks")
+        for task_idx, time_list in enumerate(time):
+            if not len(time_list) == stages[task_idx] + 1:
+                raise ValueError("time[i] should have length stages[i]")
+        for task_idx, prec_list in enumerate(prec):
+            if not len(prec_list) == stages[task_idx] + 1:
+                raise ValueError("prec[i] should have length stages[i]")
+        if not len(prec) == num_tasks:
+            raise ValueError("prec should have length num_tasks")
+        if not len(prio) == num_tasks:
+            raise ValueError("prio should have length num_tasks")
+        if not len(dead) == num_tasks:
+            raise ValueError("dead should have length num_tasks")
+
+        # delta is the basic increment of reward
+        # NOTE this is a good place to play and have fun!
+        delta = self.delta
+
+        # Compute the expected rewards for all stages of all tasks
+        # R[i][l] is the reward for completing the first l stages of task i before the deadline
+        R = []
+        for i in range(num_tasks):
+            R.append([None]*len(prec[i]))
+            for l in range(len(prec[i])):
+                reward = self.reward(prec[i][l], prio[i])
+                R[i][l] = self.quantize(reward, delta)
+
+        # TODO add a sanity check here that R[i][l] is weakly decreasing WRT l for all i
+
+        if verbose:
+            # Sanity check:
+            print("R:")
+            for i, r_list in enumerate(R):
+                line = ""
+                for j, r in enumerate(r_list):
+                    line += " "+str(r)
+                print(line)
+
+        # Find the solutions to all problems in tables S and P
+        if not self.compute_tables_from_scratch(num_tasks, stages, time, R, dead, delta):
+            print('Error in compute_tables_from_scratch')
+            exit()
+
+        if verbose:
+            # Sanity check:
+            self.printSP()
+
+        # Find optimal depths from the completed tables
+        depth_sched, reward_sched, time_sched = self.find_optimal_depths_from_tables(R, time, verbose)
+
+        # If verbose, print out the solution schedule that was found... messily for now... 
+        if verbose:
+            if depth_sched == None:
+                print("No viable schedule for these inputs")
+                return None
+            print("Depth schedule that achieves reward", sum(reward_sched), "  ")
```

Comment:
We shouldn't be printing here if this API is going to be used for an actual scheduler

Reply:
There are many print statements throughout that provide useful information from our research perspective, but I have kept all of them in "if verbose" blocks. So, when called from the scheduler, we can simply use "verbose=False" to avoid them, but this way we can keep them around for the research side of things... after all we aren't making production code, we are doing research

Reply:
listen pal... there wouldn't be any "research" if there we didn't have production code. When we finally do experiments the we will be looking at the speed of the algorithm with GPU workloads it will be a lot slower if we are printing every quantum.

Reply:
Ha, the code does not print anything as long as you call with "verbose=False"